doc/dev/osd_internals/mclock_wpq_cmp_study.rst

```diff
@@ -0,0 +1,488 @@
+8. As soon as the backfill/recovery starts, the test proceeds to initiate client
+   I/O on the client pool on another thread using a single client.
+9. During step 8 above, statistics related to the client latency and
+   bandwidth are captured by cbt. The test also captures the total number of
+   misplaced objects and the number of misplaced objects recovered per second.
+
+To summarize, the steps above creates 2 pools during the test. Recovery is
+triggered on one pool and client I/O is triggered on the other simultaneously.
+Statistics captured during the tests are discussed below.
+
+
+Non-Default Ceph Recovery Options
+`````````````````````````````````
+
+Apart from the non-default bluestore throttle already mentioned above, the
+following set of Ceph recovery related options were modified for tests with both
+the WPQ and mClock schedulers.
+
+- ``osd_max_backfills`` = 1000
+- ``osd_recovery_max_active`` = 1000
+- ``osd_async_recovery_min_cost`` = 1
+
+The above options puts a high limit on the number of concurrent local and
+remote backfill operations per OSD. It is under these conditions that the
+capabilities of the mClock scheduler is tested and discussed below.
+
+Test Results
+============
+
+Test Results With NVMe SSDs
+```````````````````````````
+
+Client Throughput Comparison
+----------------------------
+
+The chart below shows the average client throughput comparison across the
+schedulers and their respective configurations.
+
+.. image:: ../../images/mclock_wpq_study/Avg_Client_Throughput_NVMe_SSD_WPQ_vs_mClock.png
+
+
+WPQ(def) in the chart shows the average client throughput obtained
+using the WPQ scheduler with all other Ceph configuration settings set to
+default values. The default setting for ``osd_max_backfills`` limits the number
+of concurrent local and remote backfills or recoveries per OSD to 1. As a
+result, the average client throughput obtained is impressive at just over 18000
+IOPS when compared to the baseline value which is 21500 IOPS.
+
+However, with WPQ scheduler along with non-default options mentioned in section
+`Non-Default Ceph Recovery Options`_, things are quite different as shown in the
+chart for WPQ(BST). In this case, the average client throughput obtained drops
+dramatically to only 2544 IOPS. The non-default recovery options clearly had a
+significant impact on the client throughput. In other words, recovery operations
+overwhelm the client operations. Sections further below discuss the recovery
+rates under these conditions.
+
+With the non-default options, the same test was executed with mClock and with
+the default profile(*high_client_ops*) enabled. As per the profile allocation,
+the reservation goal of 50% (10750 IOPS) is being met with an average throughput
+of 11209 IOPS during the course of recovery operations. This is more than 4x
+times the throughput obtained with WPQ(BST).
+
+Similar throughput with the *balanced*(11017 IOPS) and *high_recovery_ops*
```

Comment:
I think you need a space before the ( - it's throwing off the formatting in the rendered version

Reply:
Fixed. Thanks for pointing this out.